drivers/adc/adm1177/adm1177.c

```diff
@@ -0,0 +1,316 @@
+		if(ret)
+			return ret;
+
+		return 0;
+	} else {
+		return -EINVAL;
+	}
+}
+
+/**************************************************************************//**
+ * @brief - Gets the voltage/current readback from the ADM1177s I2C ADC,
+ *	    if the device is initialized right and transmits the requested
+ *	    values.
+ *
+ * @param device - The device structure.
+ *
+ * @param conv_voltage - The variable through which the voltage read from the
+ * 			 ADC is received if requested.
+ * 			 If it is 0, it means it was not requested.
+ *
+ * @param conv_current - The variable through which the current read from the
+ * 			 ADC is received if requested.
+ * 			 If it is 0, it means it was not requested.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+******************************************************************************/
+int adm1177_readback(struct adm1177_dev *device,
+		     uint16_t* conv_voltage,
+		     uint16_t* conv_current)
+{
+	int ret;
+
+	uint8_t data[3];
+
+	switch(device->last_command) {
+	case ADM1177_CURRENT_EN:
+		ret = no_os_i2c_read(device->i2c_desc, data, 2, 1);
+		if(ret)
+			return ret;
+
+		if(conv_current == NULL)
+			return -EINVAL;
+
+		/* First byte received contains the 8 MSBs of the
+		   current and is left shifted by 4, then the second
+		   byte contains the 4 LSBs of the current and is right
+		   shifted by 4, and when the OR operation between the
+		   two bytes happen we obtain the a 12-bit value of the
+		   current. */
+		*conv_current = (data[0] << 4) | (data[1] >> 4);
+
+		return 0;
+	case ADM1177_VOLTAGE_EN:
+		ret = no_os_i2c_read(device->i2c_desc, data, 2, 1);
+		if(ret)
+			return ret;
+
+		if(conv_voltage == NULL)
+			return -EINVAL;
+
+		/* First byte received contains the 8 MSBs of the
+		   voltage and is left shifted by 4, then the second
+		   byte contains the 4 LSBs of the voltage and is right
+		   shifted by 4, and when the OR operation between the
+		   two bytes happen we obtain the a 12-bit value of
+		   the voltage. */
+		*conv_voltage = (data[0] << 4) | (data[1] >> 4);
+
+		return 0;
+	case ADM1177_VOLTAGE_AND_CURRENT_EN:
+		ret = no_os_i2c_read(device->i2c_desc, data, 3, 1);
+		if(ret)
+			return ret;
+
+		/* In the data array we obtain 3 bytes from the read,
+		   first one contains the 8 MSBs of the Voltage value,
+		   second one contains the 8 MSBs of the Current value,
+		   and the third one contains the 4 LSBs of the Voltage
+		   then the Current value, respectively.
+		   For the Voltage - By left shifting the first byte by
+		   4 we are shifting the 8 MSBs of the voltage by 4,
+		   and adding the third byte right shifted by 4 we add
+		   the remaining 4 LSBs of the voltage.
+		   For the Current -  By left shifting the second
+		   byte by 4 we are shifting the 8 MSBs of the current
+		   by 4, and adding the third byte masked by 0xF
+		   we add the remaining 4 LSBs of the current.*/
+		*conv_voltage = (data[0] << 4) | (data[2] >> 4);
```

Comment:
missing non NULL check for both conv_current and conv_voltage

Reply:
Will add.